### PR TITLE
yihan create submit for review button

### PR DESCRIPTION
**Pull request by AriaYu927**
# Description
image

## Related PRS (if any):
This frontend PR is related to the [#443 ](https://github.com/OneCommunityGlobal/HGNRest/pull/443)backend PR.

## Main changes explained:
- Update file TeamMemberTask.jsx for including ReviewButton component, and transfering userId and updateTask function to it.
- Update file TeamMemberTasks.jsx for transfering userId to TeamMemberTask
- Create file ReviewButton.jsx for introducing review button.
- Create file reviewButton.css for formatting review button

## How to test:
1. check into current branch on both frontend and backend.
2. do `npm install` and `...` to run this PR locally
3. log as any user (make sure the user has task assigned and all the accounts you use for this test are in same team)
4. go to dashboard → Tasks, verify only the user who is assigned the task can see "submit for review".
5. Click on "submit for review" -> Submit, see if the button turns to
a. "Ready for review" dropdown for Owners/Admins/Managers/Mentors who can see this user in their tasks list.
b. "More work needed, reset this button" to the user who is assigned the task (only for volunteer and core team user)
c. disabled "Ready for review" button for other type account who can see this user in their tasks list.
6. Click on "More work needed, reset this button" button, see it reset to "Submit for Review".
7.  repeat step 5.
8. go to "Ready for review" dropdown -> “More work needed, reset this button”, the dropdown and disabled button should disappear, and "submit for review" button shows again (same as step 6).
8. repeat step 5
9. go to "Ready for review" dropdown -> “✓ as complete and remove task”, see if the task is removed from task list.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/f1df014a-5a0e-4f56-9e2d-49633b1853b9

## Note:
It may take some time for button to update.

### CODE REVIEW

**Comment by AaronPersaud** [comment]
Hi, I've tested it and it works perfectly for me for admin role.

<b>Submitted review for task for Volunteer role</b>

image

However, when I log into manager/mentor roles I do not see the button for accepting/dismissing the review.

<b>Dashboard for Manager role</b>

image

Any thoughts as to why this is happening? I even added this manager to the project, but the task still isn't showing up.

**Comment by AriaYu927** [comment]
> Hi, I've tested it and it works perfectly for me for admin role.
> 
> Submitted review for task for Volunteer role
> 
> image
> 
> However, when I log into manager/mentor roles I do not see the button for accepting/dismissing the review.
> 
> Dashboard for Manager role
> 
> image
> 
> Any thoughts as to why this is happening? I even added this manager to the project, but the task still isn't showing up.

Hi @AaronPersaud , thank you for testing! To solve this issue, you have to include your mgr/mentor and volunteer account in one team. I've updated the description, please check.

**Review by xaanders — COMMENTED** [review]
Hi! I have tested your PR. Functionality works well! 
I tried different users, manager, admin, volunteer and core.

[scrnli_7_17_2023_11-57-34 AM.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108314277/2fb8123c-7559-41da-8945-32a663aaed18)
What could I see as core: 
photo_2023-07-17_12-44-16


However I have a suggestion regarding the code: 

```
  const toggle = () => {
    setModal(true);
  }
  const toggleModal = () => {
    setModal(!modal);
  };
```
It could be done as one function just checking passed arguments as they are passed inside onClick event listeners 
```
    <Button className='reviewBtn' color='primary' onClick={() => {toggle("Submitted");}}>
              Submit for Review
            </Button>
    <DropdownItem onClick={() => {toggle("Reviewed");}}>
```

**Comment by AriaYu927** [comment]
> Hi! I have tested your PR. Functionality works well! I tried different users, manager, admin, volunteer and core.
> 
> [scrnli_7_17_2023_11-57-34 AM.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108314277/2fb8123c-7559-41da-8945-32a663aaed18) What could I see as core: photo_2023-07-17_12-44-16
> 
> However I have a suggestion regarding the code of review button:
> 
> ```
>   const toggle = () => {
>     setModal(true);
>   }
>   const toggleModal = () => {
>     setModal(!modal);
>   };
> ```
> 
> It could be done as one function just checking passed arguments as they are passed inside onClick event listeners
> 
> ```
>     <Button className='reviewBtn' color='primary' onClick={() => {toggle("Submitted");}}>
>               Submit for Review
>             </Button>
>     <DropdownItem onClick={() => {toggle("Reviewed");}}>
> ```

Hi @xaanders , thanks for your testing! I've updated this PR based on your suggestion, please check if that's what you mean. Thank you!

**Review by AaronPersaud — APPROVED** [review]
Hi, I've tested and it looks good! Thank you for helping me with the issue I had yesterday!

<b>Mentor dismissing and approving review</b>

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/10265513/900f1a20-5b07-4f9a-8259-c8584661abe9

- each task has a separate button
- can cancel and close modal after clicking "Submit for review" button
- button changes after submit, and able to reset it after
- mentor/manager/admin/owner can dismiss or approve submission
  - button resets for user on dismissal
- button disappears for both parties after approving

Nice work!

**Comment by Lucas-E** [comment]
Great job everyone! Moving this one to final review!

**Comment by xaanders** [comment]
> > Hi! I have tested your PR. Functionality works well! I tried different users, manager, admin, volunteer and core.
> > [scrnli_7_17_2023_11-57-34 AM.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108314277/2fb8123c-7559-41da-8945-32a663aaed18) What could I see as core: photo_2023-07-17_12-44-16
> > However I have a suggestion regarding the code of review button:
> > ```
> >   const toggle = () => {
> >     setModal(true);
> >   }
> >   const toggleModal = () => {
> >     setModal(!modal);
> >   };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It could be done as one function just checking passed arguments as they are passed inside onClick event listeners
> > ```
> >     <Button className='reviewBtn' color='primary' onClick={() => {toggle("Submitted");}}>
> >               Submit for Review
> >             </Button>
> >     <DropdownItem onClick={() => {toggle("Reviewed");}}>
> > ```
> 
> Hi @xaanders , thanks for your testing! I've updated this PR based on your suggestion, please check if that's what you mean. Thank you!

Yes, that is what I meant. Good job!

**Review by xaanders — APPROVED** [review]

**Comment by mahamakhan** [comment]
Checked from admin and volunteer account. Nothing changes. Maybe I am doing something wrong. Can you help me with this? 



https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/106139234/e1a2ffe3-a501-4b57-975f-f35abe8f84ed

**Comment by AriaYu927** [comment]
> Checked from admin and volunteer account. Nothing changes. Maybe I am doing something wrong. Can you help me with this?
> 
>  Recording.2023-07-18.220558.mp4

Have you checked into this branch on both frontend and backend?

File: src/components/TeamMemberTasks/ReviewButton.jsx
Comment (Cirakuze) [inline src/components/TeamMemberTasks/ReviewButton.jsx:36]: Is the optional chaining necessary for `resources` here?

File: src/components/TeamMemberTasks/ReviewButton.jsx
Comment (Cirakuze) [inline src/components/TeamMemberTasks/ReviewButton.jsx:47]: I think this could be
```
newResources = {
  ...resource,
  reviewStatus: newStatus,
  completedTask: newStatus === "Reviewed",
};
```

File: src/components/TeamMemberTasks/ReviewButton.jsx
Comment (Cirakuze) [inline src/components/TeamMemberTasks/ReviewButton.jsx:120]: To improve code readability, please refactor to avoid the nested ternaries.

File: src/components/TeamMemberTasks/ReviewButton.jsx
Comment (Cirakuze) [inline src/components/TeamMemberTasks/ReviewButton.jsx:49]: Would this do the same as you intended?
```
const newResource = { ...resource };
if (resource.userID === user.personId) {
  newResources.reviewStatus = newStatus;
  newResources.completedTask = newStatus === "Reviewed";
}
return newResource;
```

Comment (AriaYu927) [reply src/components/TeamMemberTasks/ReviewButton.jsx]: updated the code based on your above three suggestions, please check.

**Comment by igorrochadasilva** [comment]
Hi @AriaYu927 !
I hope you are well = D

I accessed the volunteer's dashboard and clicked the submit button, but when I return to the dashboard/home, the page ends up not loading in its entirety because it has a very large content, so it ends up not loading for me : ( 

1013_01_dashboard-user-volunteer

1013_02_dashboard-user-volunteer

**Review by winghojackyli — DISMISSED** [review]
Hi @AriaYu927 , I've tested the PR and it worked great. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108389005/48a48d31-dbb2-400d-864a-6870a193b1cd

**Comment by AriaYu927** [comment]
> Hi @AriaYu927 ! I hope you are well = D
> 
> I accessed the volunteer's dashboard and clicked the submit button, but when I return to the dashboard/home, the page ends up not loading in its entirety because it has a very large content, so it ends up not loading for me : (
> 
> 1013_01_dashboard-user-volunteer
> 
> 1013_01_dashboard-loading-page

Hi @igorrochadasilva , thanks for your testing! This is caused by another issue and has been resolved today. Please try `npm install` then `npm run build:local` on frontend.

**Comment by Cirakuze** [comment]
Hi @AriaYu927, clicking the "Submit for Review" button triggers a pop up, but clicking "Submit" does not seem to trigger the intended outcome. Same for clicking "Ready for Review" and then either option. It does not seem to take an effect.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/15364639/da8533a4-7ef0-4d5e-bc0f-46b7c8df7b34

**Comment by AriaYu927** [comment]
> Hi @AriaYu927, clicking the "Submit for Review" button triggers a pop up, but clicking "Submit" does not seem to trigger the intended outcome. Same for clicking "Ready for Review" and then either option. It does not seem to take an effect.
> 
>  PR.1013.mov

Hi @Cirakuze , I checked both of your volunteer and admin dashboard, it seems like you submit the task successfully. 

Volunteer dashboard:
截屏2023-08-03 08 05 43

Admin dashboard:
截屏2023-08-03 08 10 05

I guess the problem is the page needs to take some time to refresh. Besides, the initial status of your button is incorrect. The "ready for review" shouldn't appear until the volunteer submit the task for review. Could you please refresh the page and test it again, and give the page enough time to refresh? Let me know if it still not work.

**Comment by Lin1404** [comment]
Hi @AriaYu927 , I've tested the PR and it worked perfect.
image
image
image

After click "reset button"
image

**Review by Lin1404 — APPROVED** [review]

**Review by zuhang2000 — COMMENTED** [review]
Hi I tested your PR following your instructions! The functionalities are working as expected!

More specifically, the following three cases are working correctly under different roles. 
a. "Ready for review" dropdown for Owners/Admins/Managers/Mentors who can see this user in their tasks list.
b. "More work needed, reset this button" to the user who is assigned the task (only for volunteer and core team user)
c. disabled "Ready for review" button for other type account who can see this user in their tasks list.

The volunteer's button gots reset correctly after Admin click "More work needed, reset this button". 
Screen Shot 2023-08-04 at 5 35 10 PM
The task is removed after Admin click "complete and remove".
Screen Shot 2023-08-04 at 5 44 46 PM

However, I do have one suggestion that when there are a lot of tasks. All those buttons look pretty messy. It would be perfect if each button can be placed right below each task instead of next to it. 
Screen Shot 2023-08-04 at 5 52 51 PM

**Comment by AriaYu927** [comment]
Hi @zuhang2000 , I've updated the PR based on your suggestion, please check. Thank you!

**Review by veronicacheng2 — APPROVED** [review]
Hi @AriaYu927, I have tested the PR and everything works as intended.

Step 1: Submit the task for review when logged in as Volunteer
image

Step 2: Button changed to Ready for Review
image

Step 3:  Mark as completed and remove task
image

Step 4: Task no longer exist on Team Member Task Tab
image

**Review by EvianTan — APPROVED** [review]

**Pull request merged**
